Page numbers not printed when using WKWebView.

rr-codes opened the pull request:
#### 2b0e8042741ba22bd408aaf764c273064647193b
<pre>
Page numbers not printed when using WKWebView.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244911">https://bugs.webkit.org/show_bug.cgi?id=244911</a>
rdar://85438043

Reviewed by Tim Horton.

In AppKit, there is an API method `drawPageBorderWithSize:` that can be
overriden to be used to arbitrarily draw contents on a page&apos;s border. In
legacy WebKit, this method was called when printing. However, in WKWebView,
the behavior differs because this method is no longer called.

This PR resolves this discrepency by ensuring this method does get called
and so the behavior remains consistent with how it works with legacy web
views.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/mac/WKPrintingView.mm:
(-[WKPrintingView drawPageBorderWithSize:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
(-[WKPrintPageBordersWebView drawPageBorderWithSize:]):
(-[WKPrintPageBordersWebView _waitUntilPrintingDone]):
(-[PrintPageBordersUIDelegate _webView:printFrame:pdfFirstPageSize:completionHandler:]):
(TEST):
(-[PrintUIDelegate callBlockAsync:]): Deleted.
(-[PrintUIDelegate _webView:printFrame:pdfFirstPageSize:completionHandler:]): Deleted.
(-[PrintUIDelegate waitForPagination]): Deleted.

Canonical link: <a href="https://commits.webkit.org/254392@main">https://commits.webkit.org/254392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b2d772ea9fcb1414f2f42d0a570c1621f3e6c9e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/88748 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/33314 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/19641 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/97954 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/154470 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/92743 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/31819 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/27438 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/80992 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/92579 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/94378 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/25241 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/75740 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25197 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80118 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80180 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68158 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/29611 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14193 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/29353 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15195 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3080 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/32785 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38119 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/31472 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34302 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -72,11 +91,37 @@ - (void)waitForPagination
+
+- (void)_webView:(WKWebView *)webView printFrame:(_WKFrameHandle *)frame pdfFirstPageSize:(CGSize)size completionHandler:(void (^)(void))completionHandler
+{
+    NSPrintInfo *pi = [[NSPrintInfo alloc] init];
```
Nit: `pi` -> `printInfo`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -86,3 +131,18 @@ - (void)waitForPagination
+    auto delegate = adoptNS([PrintPageBordersUIDelegate new]);
+    [webView setUIDelegate:delegate.get()];
+
+    NSURLRequest *request = [NSURLRequest requestWithURL:[[NSBundle mainBundle]URLForResource:@"simple" withExtension:@"html" subdirectory:@"TestWebKitAPI.resources"]];
```
Spacing: `[NSBundle mainBundle]URLForResource:` -> `[NSBundle mainBundle] URLForResource:`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -86,3 +131,18 @@ - (void)waitForPagination
+TEST(Printing, PrintPageBorders)
+{
+    auto configuration = adoptNS([WKWebViewConfiguration new]);
+    auto webView = adoptNS([[WKPrintPageBordersWebView alloc]initWithFrame:NSMakeRect(0, 0, 800, 600) configuration:configuration.get()]);
```
Spacing: `[[WKPrintPageBordersWebView alloc]initWithFrame` -> `[[WKPrintPageBordersWebView alloc] initWithFrame`.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -72,11 +91,37 @@ - (void)waitForPagination
+@end
+
+@implementation PrintPageBordersUIDelegate {
+    bool _isDone;
```
This variable is unused.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -34,13 +34,32 @@
 @end
 
-@implementation PrintUIDelegate {
+@implementation PrintWithDelayedCompletionUIDelegate {
```
Maybe `PrintWithSimulatedPageComputationUIDelegate`? Printing is async, so "delayed completion" isn't really specific to what this class does.

rr-codes replied:
this is the current class that was there for the previous test, just renamed (the new name is based on the existing test name). Should I still rename it to `PrintWithSimulatedPageComputationUIDelegate`?

pxlcoder replied:
I think so, it better distinguishes the two delegates: one that actually presents a print panel (the new one you added), the other that tests an internal method (this existing class).

pxlcoder replied:
Could also consider one class, with an initializer to choose between the two cases, but not too important since this is testing only code.

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -72,11 +91,37 @@ - (void)waitForPagination
 
 @end
 
+@interface PrintPageBordersUIDelegate : NSObject <WKUIDelegate>
```
Maybe `PrintShowingPrintPanelUIDelegate`?

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -34,13 +34,32 @@
+@end
+
+@implementation WKPrintPageBordersWebView {
+    bool _done;
```
`_didDrawPageBorder`?

alanbaradlay commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -9179,6 +9179,13 @@ void WebPageProxy::drawFooterForPrinting(WebFrameProxy& frame, FloatRect&& rect)
     m_uiClient->drawFooter(*this, frame, WTFMove(rect));
 }
 
+void WebPageProxy::drawPageBorderForPrinting(WebFrameProxy& frame, WebCore::FloatSize&& size)
```
Any particular reason for using rvalue reference here?

alanbaradlay replied:
oh nvm, it looks like all the drawing functions do that.

hortont424 replied:
Still seems bizarre

pxlcoder commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -34,13 +34,32 @@
+    _done = true;
+}
+
+- (void)_waitUntilPrintingDone
```
`_waitUntilPageBorderDrawn`?

hortont424 commented on Source/WebKit/UIProcess/mac/WKPrintingView.mm:
```diff
@@ -677,6 +677,10 @@ - (void)drawPageBorderWithSize:(NSSize)borderSize
     [currentContext restoreGraphicsState];
+    
+    [currentContext saveGraphicsState];
+    _webFrame->page()->drawPageBorderForPrinting(*_webFrame, (WebCore::FloatSize)borderSize);
```
I wonder if you need to clip to the page rect, or if something else did that for you (or maybe the page rect == the context bounds?)

hortont424 replied:
also, static_cast<WebCore::FloatSize>(...)

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -86,3 +86,63 @@ - (void)waitForPagination
 }
+
+#if PLATFORM(MAC)
+@interface WKPrintPageBordersWebView: TestWKWebView
```
Spaces on both sides of the `:`

hortont424 commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
```diff
@@ -86,3 +86,63 @@ - (void)waitForPagination
+
+- (void)_webView:(WKWebView *)webView printFrame:(_WKFrameHandle *)frame pdfFirstPageSize:(CGSize)size completionHandler:(void (^)(void))completionHandler
+{
+    NSPrintInfo *printInfo = [[NSPrintInfo alloc] init];
```
This technically leaks. It's fine since we run tests singly, but would be nice if you'd adopt it into a RetainPtr instead.

rr-codes committed:
Page numbers not printed when using WKWebView.
https://bugs.webkit.org/show_bug.cgi?id=244911
rdar://85438043

Reviewed by Tim Horton.

In AppKit, there is an API method `drawPageBorderWithSize:` that can be
overriden to be used to arbitrarily draw contents on a page's border. In
legacy WebKit, this method was called when printing. However, in WKWebView,
the behavior differs because this method is no longer called.

This PR resolves this discrepency by ensuring this method does get called
and so the behavior remains consistent with how it works with legacy web
views.

* Source/WebKit/UIProcess/PageClient.h:
(WebKit::PageClient::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.h:
* Source/WebKit/UIProcess/mac/PageClientImplMac.mm:
(WebKit::PageClientImpl::drawPageBorderForPrinting):
* Source/WebKit/UIProcess/mac/WKPrintingView.mm:
(-[WKPrintingView drawPageBorderWithSize:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKPrinting.mm:
(-[WKPrintPageBordersWebView drawPageBorderWithSize:]):
(-[WKPrintPageBordersWebView _waitUntilPrintingDone]):
(-[PrintPageBordersUIDelegate _webView:printFrame:pdfFirstPageSize:completionHandler:]):
(TEST):
(-[PrintUIDelegate callBlockAsync:]): Deleted.
(-[PrintUIDelegate _webView:printFrame:pdfFirstPageSize:completionHandler:]): Deleted.
(-[PrintUIDelegate waitForPagination]): Deleted.

Canonical link: https://commits.webkit.org/254392@main
